cpp/include/raft/buffer_base.hpp

```diff
@@ -0,0 +1,183 @@
+              size_type n = 0)
+    : _size(n),
+      _capacity(n),
+      _data(nullptr),
+      _stream(stream),
+      _allocator(allocator) {
+    if (_capacity > 0) {
+      _data = static_cast<value_type*>(
+        _allocator->allocate(_capacity * sizeof(value_type), _stream));
+      CUDA_CHECK(cudaStreamSynchronize(_stream));
+    }
+  }
+
+  ~buffer_base() {
+    if (nullptr != _data) {
+      _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+    }
+  }
+
+  value_type* data() { return _data; }
+
+  const value_type* data() const { return _data; }
+
+  size_type size() const { return _size; }
+
+  void clear() { _size = 0; }
+
+  iterator begin() { return _data; }
+
+  const_iterator begin() const { return _data; }
+
+  iterator end() { return _data + _size; }
+
+  const_iterator end() const { return _data + _size; }
+
+  /**
+   * @brief Reserve new memory size for this buffer.
+   *
+   * It re-allocates a fresh buffer if the new requested capacity is more than
+   * the current one, copies the old buffer contents to this new buffer and
+   * removes the old one.
+   *
+   * @param[in] new_capacity new capacity (in number of elements)
+   * @param[in] stream       cuda stream where allocation operations are queued
+   */
+  void reserve(const size_type new_capacity, cudaStream_t stream) {
+    set_stream(stream);
+    if (new_capacity > _capacity) {
+      value_type* new_data = static_cast<value_type*>(
+        _allocator->allocate(new_capacity * sizeof(value_type), _stream));
+      if (_size > 0) {
+        CUDA_CHECK(cudaMemcpyAsync(new_data, _data, _size * sizeof(value_type),
+                                   cudaMemcpyDefault, _stream));
+      }
+      if (nullptr != _data) {
+        _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+      }
+      _data = new_data;
+      _capacity = new_capacity;
+    }
+  }
+
+  /**
+   * @brief Resize the underlying buffer (uses `reserve` method internally)
+   *
+   * @param[in] new_size new buffer size
+   * @param[in] stream   cuda stream where the work will be queued
+   */
+  void resize(const size_type new_size, cudaStream_t stream) {
+    reserve(new_size, stream);
+    _size = new_size;
+  }
+
+  /**
+   * @brief Deletes the underlying buffer
+   *
+   * If this method is not explicitly called, it will be during the destructor
+   *
+   * @param[in] stream   cuda stream where the work will be queued
+   */
+  void release(cudaStream_t stream) {
+    set_stream(stream);
+    if (nullptr != _data) {
+      _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+    }
+    _data = nullptr;
+    _capacity = 0;
+    _size = 0;
+  }
+
+  /**
+   * @brief returns the underlying allocator used
+   *
+   * @return the allocator pointer
+   */
+  std::shared_ptr<Allocator> get_allocator() const { return _allocator; }
+
+ protected:
+  value_type* _data;
+
+ private:
+  size_type _size;
```

Comment:
Did we switch from trailing _ to preceding _?

I was also using preceding _, but after seeing

https://stackoverflow.com/questions/3136594/naming-convention-underscore-in-c-and-c-sharp-variables/20496955#20496955

saying

```
Please DO NOT use UNDERSCORES before any variable name or parameter name in C++!!!

Names beginning with an underscore or a double underscore are RESERVED for the C++ implementers. Names with an underscore are reserved for the library to work.

If you have a read at the C++ Coding Standard, you will see that in the very first page it says:
```

I switched to trailing _.

(and both Google C++ coding style guide (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) and Herb Sutter & Andrei Alexandrescu (big names in the C++ world, https://www.drdobbs.com/c-coding-standards/184401863) recommend using trailing _.)

Any reason to go for preceding _ over trailing _ considering all the above recommendations?

Reply:
Ouch... another one slipped by me due to simply copy-pasting stuff from cuML! Also a bit surprised that clang-tidy didn't catch this one! Let me see what's going on. Thanks for catching this.

Reply:
wait a min... this file has been renamed to `cpp/include/mr/buffer_base.hpp`. It has all the correct variable namings. I think you were looking at an older version of this file where the namings were done incorrectly.